core/src/main/java/io/grpc/internal/KeepAliveManager.java

```diff
@@ -0,0 +1,223 @@
+ * met:
+ *
+ *    * Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ *    * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *
+ *    * Neither the name of Google Inc. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.grpc.internal;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+
+import io.grpc.Status;
+
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Manages keepalive pings.
+ */
+public class KeepAliveManager {
+  private static final SystemTicker SYSTEM_TICKER = new SystemTicker();
+  private static long KEEPALIVE_DELAY = TimeUnit.MINUTES.toNanos(1);
+  private static long KEEPALIVE_TIMEOUT = TimeUnit.MINUTES.toNanos(2);
+  private final ScheduledExecutorService scheduler;
+  private final ManagedClientTransport transport;
+  private final Ticker ticker;
+  private State state = State.IDLE;
+  private long nextKeepaliveTime;
+  private ScheduledFuture<?> shutdownFuture;
+  private ScheduledFuture<?> pingFuture;
+  private final Runnable shutdown = new Runnable() {
+    @Override
+    public synchronized void run() {
+      if (state != State.DISCONNECTED) {
+        // We haven't received a ping response within the timeout. The connection is likely gone
+        // already. Shutdown the transport and fail all existing rpcs.
+        state = State.DISCONNECTED;
+        transport.shutdownNow(Status.UNAVAILABLE);
+      }
+    }
+  };
+  private final Runnable sendPing = new Runnable() {
+    @Override
+    public synchronized void run() {
+      if (state == State.PING_SCHEDULED
+          || (state == State.PING_DELAYED && nextKeepaliveTime <= ticker.read())) {
```

Comment:
This condition (`state == PING_DELAYED && ...`) isn't probably very important. It seems like it would be true very rarely, and if it is true we could have just handled PING_DELAYED like normal and then the scheduler will call this method again almost immediately.